application/src/main/kotlin/net/corda/v5/application/messaging/FlowSession.kt

```diff
@@ -31,6 +37,23 @@ interface FlowSession {
      */
     val counterparty: MemberX500Name
 
+    /**
+     * Session local [FlowContextProperties].
+     *
+     * If this session is part of an initiating flow, i.e. was obtained from [FlowMessaging] then this is a read only
+     * set of context properties which will be used to determine context on the initiated side. Modifying this set is
+     * only possible when session are initiated, see [FlowMessaging].
+     *
+     * If this session was passed to an initiated flow by Corda, the context properties are associated with the
+     * initiating flow at the other end of the connection. They differ from the [FlowContextProperties] available to all
+     * flows via the [FlowEngine] in that they are not a description of the context of the currently executing flow, but
+     * instead the flow which initiated it.
+     *
+     * Any calls to modify these contextProperties will throw a [CordaRuntimeException], they should be considered
+     * immutable.
+     */
+    val contextProperties: FlowContextProperties
```

Comment:
Mmmmm, so if `FlowContextProperties` should only be modified in `FlowContextPropertiesBuilder`, should we change things a bit so `FlowContextProperties` cannot be modified at all by users? 

Hopefully, this isn't treading on things we already talked about since we discussed having a read-only version before hand. 

However, with the introduction of the builder, we can have a read only version without needing changes to `FlowSession`.

If `FlowContextPropertiesBuilder` was actually a builder, then that can be where it is mutable and then after calling a `.build` method it becomes a `FlowContextProperties` which only has a `get`. Annoyingly, the naming issue becomes if `FlowContextPropertiesBuilder` is a real builder, then the functional interface you current have named `FlowContextPropertiesBuilder` would need a different name itself, and needs to pass the "new" `FlowContextPropertiesBuilder` into its `apply` method.

Doing something like that means you can remove the `Any calls to modify these contextProperties will throw a [CordaRuntimeException], they should be considered immutable.` clause you have here.

But, as always, assuming what I've said makes any sort of sense, dealing with the naming problem here doesn't look simple...

Reply:
Happy to discuss this if needed @simon-johnson-r3.